File: sdks/csharp/src/Verdocs.Sdk/Utils/Strings.cs
using System.Security.Cryptography;
using System.Text;

namespace Verdocs.Utils;

/// <summary>String helpers (js-sdk: Utils/Strings.ts).</summary>
public static class Strings
{
    // The digits Math.random().toString(36) can produce; RandomString draws from the same
    // alphabet so its output is interchangeable with the js-sdk's.
    private const string Base36Alphabet = "0123456789abcdefghijklmnopqrstuvwxyz";

    /// <summary>Returns the string with its first letter uppercased.</summary>
    /// <param name="value">The string to capitalize. May be empty.</param>
    /// <returns>The input with the first character uppercased, other characters untouched.</returns>
    public static string Capitalize(string value)
    {
        ArgumentNullException.ThrowIfNull(value);
        return value.Length == 0 ? value : char.ToUpperInvariant(value[0]) + value[1..];
    }

    /// <summary>
    /// Converts a phone-number-like string to E.164 format, assuming US numbers. A value
    /// already prefixed with "+" is only trimmed: the caller set a country code deliberately,
    /// so non-US numbers pass through untouched. Everything else has punctuation stripped,
    /// one leading zero removed, and "+1" prepended (the js-sdk assumes US for unprefixed
    /// input, and we keep that).
    ///
    /// <example>
    /// <code>
    /// Strings.ConvertToE164("(212) 555-1212"); // "+12125551212"
    /// Strings.ConvertToE164("+46766861004");   // "+46766861004"
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="input">The phone-number-like input; null is treated as empty.</param>
    /// <returns>The E.164 string, or the trimmed input when it was blank or "+"-prefixed.</returns>
    public static string ConvertToE164(string? input)
    {
        var trimmed = (input ?? string.Empty).Trim();
        if (trimmed.Length == 0 || trimmed.StartsWith('+'))
        {
            return trimmed;
        }

        var digits = new StringBuilder(trimmed.Length);
        foreach (var ch in trimmed)
        {
            if (char.IsAsciiDigit(ch))
            {
                digits.Append(ch);
            }
        }

        // One leading zero comes off after the punctuation strip because it may hide inside
        // it, e.g. "(05..." as well as "0(5...".
        var number = digits.ToString();
        if (number.StartsWith('0'))
        {
            number = number[1..];
        }

        return "+1" + number;
    }

    /// <summary>
    /// Generates a random lowercase-alphanumeric string of the given length. Adapted from
    /// the js-sdk, which slices Math.random().toString(36) and, due to an off-by-one in its
    /// substring bounds, actually returns length + 1 characters. We draw exactly
    /// <paramref name="length"/> characters from the same base36 alphabet using a
    /// cryptographic RNG (per the true-up handoff), so the output is also safe for more than
    /// DOM-id duty.
    /// </summary>
    /// <param name="length">Number of characters to generate.</param>
    /// <returns>A random string of exactly <paramref name="length"/> characters from [0-9a-z].</returns>
    public static string RandomString(int length)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(length);
        return RandomNumberGenerator.GetString(Base36Alphabet, length);
    }
}

File: sdks/python/src/verdocs/utils/strings.py
"""String helpers (js-sdk: Utils/Strings.ts)."""

from __future__ import annotations

import re
import secrets

# The digits Math.random().toString(36) can produce; random_string draws from
# the same alphabet so its output is interchangeable with the js-sdk's.
_BASE36_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyz"


def capitalize(value: str) -> str:
    """Return the string with its first letter uppercased.

    Args:
        value: The string to capitalize. May be empty.

    Returns:
        The input with the first character uppercased, other characters untouched.

    @sdkOperation string.capitalize
    @sdkGroup String
    @sdkPage Helpers
    """
    return value[:1].upper() + value[1:]


def convert_to_e164(value: str | None) -> str:
    """Convert a phone-number-like string to E.164 format, assuming US numbers.

    A value already prefixed with "+" is only trimmed: the caller set a country
    code deliberately, so non-US numbers pass through untouched. Everything
    else has punctuation stripped, one leading zero removed, and "+1" prepended
    (the js-sdk assumes US for unprefixed input, and we keep that).

    Example:
        convert_to_e164("(212) 555-1212")  # "+12125551212"
        convert_to_e164("+46766861004")    # "+46766861004"

    Args:
        value: The phone-number-like input; None is treated as empty.

    Returns:
        The E.164 string, or the trimmed input when it was blank or "+"-prefixed.

    @sdkOperation string.convertToE164
    @sdkGroup String
    @sdkPage Helpers
    """
    number = (value or "").strip()
    if not number or number.startswith("+"):
        return number

    number = re.sub(r"[^0-9]", "", number)
    # One leading zero comes off after the punctuation strip because it may
    # hide inside it, e.g. "(05..." as well as "0(5...".
    number = re.sub(r"^0", "", number)
    return f"+1{number}"


def random_string(length: int) -> str:
    """Generate a random lowercase-alphanumeric string of the given length.

    Adapted from the js-sdk, which slices Math.random().toString(36) and, due
    to an off-by-one in its substring bounds, actually returns length + 1
    characters. We draw exactly `length` characters from the same base36
    alphabet using the secrets module (per the true-up handoff), so the output
    is also safe for more than DOM-id duty.

    Args:
        length: Number of characters to generate.

    Returns:
        A random string of exactly `length` characters from [0-9a-z].

    @sdkOperation string.randomString
    @sdkGroup String
    @sdkPage Helpers
    """
    return "".join(secrets.choice(_BASE36_ALPHABET) for _ in range(length))
